plumbing/format/gitignore/dir.go

```diff
@@ -61,7 +61,9 @@ func ReadPatterns(fs billy.Filesystem, path []string) (ps []Pattern, err error)
 	for _, fi := range fis {
 		if fi.IsDir() && fi.Name() != gitDir {
 			var subps []Pattern
-			subps, err = ReadPatterns(fs, append(path, fi.Name()))
+			tempPath := make([]string, len(path))
+			copy(tempPath, path)
+			subps, err = ReadPatterns(fs, append(tempPath, fi.Name()))
```

Comment:
Think it would be cleaner to have the copy done by ParsePattern(), the object that takes ownership of the slice. Or in the readIgnoreFile func to share ownership. Here you would still get bugs if calling ReadPatterns incorrectly, e.g:

```go
path := []string{"a", "b"}
ps, _ := ReadPatterns(fs, path)
path[1] = "change" // ps now corrupted
```